### PR TITLE
feat(cli/lint): add the `--rule` option

### LINKED ISSUE #58: 📎 Implement `biome lint --rule=<rule-name>`
### Description

Executing `biome lint --rule=<rule-name>` should only run the rule with name `<rule-name>`, otherwise it will throw an error. See [rome#4747](https://github.com/rome/tools/discussions/4747#discussioncomment-6597436) for more context.

EDIT: `--rule=<rule-name>` should override the config, i.e. run `<rule-name>` even if the rule is disabled in the config. However, the rule options and severity level should be taken into account.

If the option is combined with `--summary` (e.g. `biome lint --summary --rule=<rule-name>`) we should restrict the summary to `<rule-name>`. See #57 for more details about the `--summary` option.

An example:

```
$ biome lint --rule useSingleCaseStatement src

src/services/types.ts:1121:5 lint/nursery/useNamingConvention ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ⚠ This property name should be in camelCase.
  
  > 1121 │     InsertSpaceAfterConstructor?: boolean;
         │     ^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  ℹ The name could be renamed to `insertSpaceAfterConstructor`.


$ biome lint --summary --rule useSingleCaseStatement src

Rule name                          Diagnostics
nursery/useSingleCaseStatement        1
```

### Possible extensions

- Multi rule filtering (keeping diagnostics of a given set of rules)
- reverse filtering, i.e. keeping all diagnostics except the diagnostics of a given rule

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

Fix #58

This PR adds a new option `--rule` on the `biome lint` command.
The option allows executing a given rule on a set of files.
The configuration is taken into account.
Only the severity is overridden to its default, i.e. `error` for recommended rules and `warn` for other rules.
This ensures that the rule is not `off`.
The rule is executed even if `recommended` is disabled.
However, **the rule is not executed if the linter is disabled in the configuration.**

To implement this feature, I introduced a new type `RuleCode` that references a rule with static string slices.
This allows passing the rule `group/name` without cloning it.
I modified `Rules::matches_diagnostic_code` to return static string slices. It acts as a static string interner.
I also added methods to set the severity level of a given rule.

The argument of the rule option is passed as parameters in the different calls. 

## Test Plan

I added tests.

### CODE REVIEW

File: crates/biome_lsp/src/session.rs [hunk 1]
```diff
@@ -311,6 +311,7 @@ impl Session {
                 path: biome_path.clone(),
                 categories,
                 max_diagnostics: u64::MAX,
+                rule: None,
```
Comment (Conaclos) [inline crates/biome_lsp/src/session.rs:314]: Here I set `rule` to `None` because I do not have access to this data. I am not sure about the consequences. Maybe it only affects the LSP. Since the CLI is the only one using the `rule` option, it's probably OK.

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,34 @@ our [guidelines for writing a good changelog entry](https://github.com/biomejs/b
...
+  ```
+
+  The option takes the rule configuration into account.
+
+  The severity level of the executed rule is always set to its default,
+  i.e. `error` for a recommended rule or `warn` otherwise.
+
+  The option is compatible with other options such as `--apply` and `--reporter`.
+
+  Contributed by @Conaclos
```
Comment (ematipico) [inline CHANGELOG.md:44]:
``````suggestion
- Add a new option `--rule` to the command `biome lint` ([#58](https://github.com/biomejs/biome/issues/58)).

  This new option allows you to execute a single rule. This option is convenient to test a rule or apply the code fixes of a single rule.

  For example, you can execute the `style/useNamingConvention` rule on a set of files:

  ```shell
  biome lint --rule=style/useNamingConvention src/index.js src/main.js
  ```

  If the rule has code action (autofix), you can use `--apply` to apply the fix:

  ```shell
  biome lint --rule=style/useNamingConvention --apply index.js src/main.js
  ```

  The option takes the rule configuration into account.

  The severity level of the executed rule is always set to its default,
  i.e. `error` for a recommended rule or `warn` otherwise.

  The option is compatible with other options such as `--apply`, `--apply-unsafe` and `--reporter`.

  Contributed by @Conaclos
``````

>  The option takes the rule configuration into account.

You might want to rephrase this. I can't suggest something because I don't know what it means, sorry about that.

File: crates/biome_cli/src/execute/mod.rs [hunk 3]
```diff
@@ -125,6 +126,8 @@ pub enum TraversalMode {
         /// 1. The virtual path to the file
         /// 2. The content of the file
         stdin: Option<Stdin>,
+
+        rule: Option<RuleCode>,
```
Comment (ematipico) [inline crates/biome_cli/src/execute/mod.rs:130]: Doc comment missing

File: crates/biome_configuration/src/linter/mod.rs [hunk 4]
```diff
@@ -219,3 +220,65 @@ pub struct RuleWithOptions<T: Default> {
...
+pub struct RuleCode {
+    pub group: &'static str,
+    pub name: &'static str,
+}
+
+impl<'a> From<RuleCode> for RuleFilter<'a> {
+    fn from(RuleCode { group, name }: RuleCode) -> Self {
+        RuleFilter::Rule(group, name)
+    }
+}
```
Comment (ematipico) [inline crates/biome_configuration/src/linter/mod.rs:234]: Have you considered making `name` a `Option<&'static str>`. This would allow us to create a `RuleFilter::Group` and run also rules that belong to a group:

```
biome lint --rule=complexity
```

File: crates/biome_configuration/src/linter/mod.rs [hunk 5]
```diff
@@ -219,3 +220,65 @@ pub struct RuleWithOptions<T: Default> {
...
+    type Err = &'static str;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Some((group, rule_name)) = s.split_once('/') {
+            if let Some((group, name)) = Rules::matches_diagnostic_code(group, rule_name) {
+                Ok(RuleCode { group, name })
+            } else {
+                Err("the rule doesn't exist.")
+            }
+        } else {
+            Err("a group and a rule name separated by a slash is required.")
```
Comment (ematipico) [inline crates/biome_configuration/src/linter/mod.rs:246]: Did you start the phrase with a lower case letter on purpose? e.g. `the rule...` VS `The rule...`

File: crates/biome_configuration/src/linter/rules.rs [hunk 6]
```diff
@@ -67,24 +67,28 @@ impl DeserializableValidator for Rules {
...
-                "performance" => Performance::has_rule(rule_name).then_some((category, rule_name)),
-                "security" => Security::has_rule(rule_name).then_some((category, rule_name)),
-                "style" => Style::has_rule(rule_name).then_some((category, rule_name)),
-                "suspicious" => Suspicious::has_rule(rule_name).then_some((category, rule_name)),
-                _ => None,
-            },
+    #[doc = r" Usually the code is built like {group}/{rule_name}"]
+    pub fn matches_diagnostic_code(
+        group: &str,
+        rule_name: &str,
```
Comment (ematipico) [inline crates/biome_configuration/src/linter/rules.rs:73]: Why was this changed?

File: crates/biome_configuration/src/linter/rules.rs [hunk 7]
```diff
@@ -910,8 +999,8 @@ impl A11y {
         index_set
     }
     #[doc = r" Checks if, given a rule name, matches one of the rules contained in this category"]
-    pub(crate) fn has_rule(rule_name: &str) -> bool {
-        Self::GROUP_RULES.contains(&rule_name)
+    pub(crate) fn has_rule(rule_name: &str) -> Option<&'static str> {
+        Some(Self::GROUP_RULES[Self::GROUP_RULES.binary_search(&rule_name).ok()?])
```
Comment (ematipico) [inline crates/biome_configuration/src/linter/rules.rs:1003]: What measures do we have to ensure that groups stay sorted?

File: crates/biome_service/src/file_handlers/json.rs [hunk 8]
```diff
@@ -290,13 +290,23 @@ fn lint(params: LintParams) -> LintResults {
...
-                .unwrap_or_default()
-                .into_iter()
-                .collect::<Vec<_>>();
+            let mut rules = settings.as_rules(params.path.as_path());
+            let rule_filter_list = if let Some(rule) = params.rule {
+                // We execute a single rule because the `--rule` filter is specified.
+                // Set the severity level to its default.
+                if let Some(rules) = rules.as_mut() {
+                    rules.to_mut().set_default_severity(rule.group, rule.name);
+                }
```
Comment (ematipico) [inline crates/biome_service/src/file_handlers/json.rs:299]: If we take receive a `&mut` using `as_mut`, why do we need to call `to_mut`?

File: crates/biome_service/src/file_handlers/javascript.rs [hunk 9]
```diff
@@ -313,24 +313,39 @@ pub(crate) fn lint(params: LintParams) -> LintResults {
...
+                    "correctness",
+                    "noDuplicatePrivateClassMembers",
+                ));
+                rule_filter_list.push(RuleFilter::Rule("correctness", "noInitializerWithDefinite"));
+                rule_filter_list.push(RuleFilter::Rule("correctness", "noSuperWithoutExtends"));
+                rule_filter_list.push(RuleFilter::Rule("nursery", "noSuperWithoutExtends"));
+                if let Some(rule) = params.rule {
+                    rule_filter_list.clear();
+                    rule_filter_list.push(rule.into())
+                }
```
Comment (ematipico) [inline crates/biome_service/src/file_handlers/javascript.rs:346]: Is this correct? We already check `params.rule` before

**Review by ematipico — COMMENTED** [review]
A believe we can make this option more powerful by running rules that belong to a single group too.

File: crates/biome_configuration/src/linter/mod.rs (hunk 4 above)
Comment (Conaclos) [reply crates/biome_configuration/src/linter/mod.rs:234]: Yeah, I considered that. However, I gave up the idea because I am still unsure how we will evolve the way we refer to rules. We previously talked about removing the group from the rule name, and we also talked about rule domains or languages.

However, with a second thought, I accept that even if we change the way we refer to rules, it may still make sense to enable a particular group's rule.

Do you think we should rename the option? `--rule-filter`, `--filter-rule`, `--rules`?

File: crates/biome_configuration/src/linter/mod.rs (hunk 5 above)
Comment (Conaclos) [reply crates/biome_configuration/src/linter/mod.rs:246]: Because it is displayed after a double colon `:` in bpaf. I didn't check if we use another convention in other errors.

File: crates/biome_configuration/src/linter/rules.rs (hunk 6 above)
Comment (Conaclos) [reply crates/biome_configuration/src/linter/rules.rs:73]: Because I need static string slices to avoid copying the rule code over and over. Previously the emitted string slices had a lifetime that depends on the function parameters.

File: crates/biome_configuration/src/linter/rules.rs (hunk 7 above)
Comment (Conaclos) [reply crates/biome_configuration/src/linter/rules.rs:1003]: Good point. It is something I forgot to check.

File: crates/biome_configuration/src/linter/mod.rs (hunk 4 above)
Comment (ematipico) [reply crates/biome_configuration/src/linter/mod.rs:234]: > Yeah, I considered that. However, I gave up the idea because I am still unsure how we will evolve the way we refer to rules. We previously talked about removing the group from the rule name, and we also talked about rule domains or languages.

I know. However, since things aren't final and we haven't taken any concrete steps towards this, we shouldn't hold ourselves from shipping new features.

> Do you think we should rename the option? `--rule-filter`, `--filter-rule`, `--rules`?

I fear that the word `filter` and the plural word `rules` might give an incorrect understanding and imply the possibility of accepting **multiple** rules/filters.  However, I think `--rule-filter` can work. 

Another suggestion: we could consider a `--group` option, and if both `--group` and `--rule` are passed, we can emit an "incompatible arguments" diagnostic.

File: crates/biome_service/src/file_handlers/json.rs (hunk 8 above)
Comment (ematipico) [reply crates/biome_service/src/file_handlers/json.rs:299]: Thank you for the answer. Would you mind going into more details? I am genuinely curious about the technical details 😄

File: crates/biome_configuration/src/linter/mod.rs (hunk 4 above)
Comment (Conaclos) [reply crates/biome_configuration/src/linter/mod.rs:234]: Or `--only`? Maybe it is too generic? Otherwise, I can extend the option and accept multiple rules?

Also, do you think that a group should enable all rules of the group, even the ones disabled in the configuration/disabled by default? I am more inclined to think we should respect the severity level of every rule when a group is passed?

File: crates/biome_service/src/file_handlers/json.rs (hunk 8 above)
Comment (Conaclos) [reply crates/biome_service/src/file_handlers/json.rs:299]: In the same function, we compute the final `rules` configuration by taking `overrides` into account. To avoid unnecessary allocation, this computation uses a copy-on-write strategy. Thus, it produces a `Cow<Rules>`.

In the present code we want to override the severity level of the executed rule. We have to mutate `rules`.
We call `to_mut` to turn the Cow into a `Cow::Owned` (if it is not already the case) and get a mutable reference to it.

We could avoid some allocation by also using a copy-on-write strategy. This could require to generate a `set_default_severity` that takes a `Cow<Rules>` instead of `&mut self`.

File: crates/biome_configuration/src/linter/mod.rs (hunk 4 above)
Comment (ematipico) [reply crates/biome_configuration/src/linter/mod.rs:234]: > Otherwise, I can extend the option and accept multiple rules?

I really depends on the scope of the feature and what you envisioned. I don't mind also accepting multiple rule filters. However, I wouldn't expose the term "filter" to the user because it's an internal terminology. Users know them as rules (and maybe groups).

> Also, do you think that a group should enable all rules of the group, even the ones disabled in the configuration/disabled by default? I am more inclined to think we should respect the severity level of every rule when a group is passed?

Yes, it makes sense to respect the configuration.

File: crates/biome_configuration/src/linter/rules.rs (hunk 7 above)
Comment (Conaclos) [reply crates/biome_configuration/src/linter/rules.rs:1043]: I added tests at the end of the file.

File: crates/biome_cli/src/commands/mod.rs [hunk 10]
```diff
@@ -151,6 +152,12 @@ pub enum BiomeCommand {
 
         #[bpaf(external, hide_usage)]
         cli_options: CliOptions,
+
+        /// Run only the given rule taking into account the options set in the configurations files.
+        /// The severity level of the rule is set to its default.
+        #[bpaf(long("rule"), argument("NAME"))]
```
Comment (ematipico) [inline crates/biome_cli/src/commands/mod.rs:158]: - we could add two examples, one for executing a rule, one for executing a group. This will appear in the help section and reference page
- instead of `NAME` we could have `GROUP|RULE_NAME` (not that great as a suggestion though, feel free to discard it)

File: CHANGELOG.md [hunk 11]
```diff
@@ -15,6 +15,41 @@ our [guidelines for writing a good changelog entry](https://github.com/biomejs/b
...
+#### New features
+
+- Add a new option `--rule` to the command `biome lint` ([#58](https://github.com/biomejs/biome/issues/58)).
+
+  This new option allows you to execute a single rule. This option is convenient to test a rule or apply the code fixes of a single rule.
+
+  For example, you can execute the `style/useNamingConvention` rule on a set of files:
+
+  ```shell
+  biome lint --rule=style/useNamingConvention src/main.js
```
Comment (ematipico) [inline CHANGELOG.md:27]: The reason why I suggested two files, is because the previous sentence says "set of files", hence the example shows a set of files.

I suggest reviewing the wording.

**Comment by ematipico** [comment]
@Conaclos you forgot to address

> taking the configurations file into account

It's still somewhere inside the doc comments and in the description of the PR. Could you please make sure to follow up? I still find that phrase opaque. I truly don't understand what it means 😅

**Comment by Conaclos** [comment]
> It's still somewhere inside the doc comments and in the description of the PR. Could you please make sure to follow up? I still find that phrase opaque. I truly don't understand what it means 😅

I made the following changes;

> Run only the given rule taking the configurations into account

Then I changed to:

> Run only the given rule taking the rule options in the configuration file into account

And so:

> Run only the given rule or rule group taking the configuration file into account

I am unsure how to address your concern
